torch/distributed/distributed_c10d.py

```diff
@@ -1412,47 +1480,99 @@ def recv(tensor: torch.Tensor, src: Optional[int] = None, group: Optional[Proces
-        _check_op(op)
-        _check_single_tensor(tensor, "tensor")
-        return object.__new__(cls)
+    def wait(self):
+        for work in self.works:
+            work.wait()
 
 
 @contextlib.contextmanager
-def _coalescing_manager(group, device, reqs):
-    if group is None:
-        group = _get_default_group()
-    group._start_coalescing(device)
+def _coalescing_manager(
+    group: Optional[ProcessGroup] = None,
+    device: Optional[torch.device] = None,
+    async_ops: Optional[bool] = False,
+):
+    """
+    A context manager used to coalesce collectives or P2P operations when possible.
+
+    Args:
+        group (`ProcessGroup`, optional): The process group to work on. If None,
+            the default process group will be used.
+        device (`torch.device`, optional): Default is None, set to a device if
+            there isn't a `**_coalesced` implementation by the backend.
+        async_ops (`bool`, optional): whether the coalesced ops are async ops.
+
+    Examples:
+        >>> # xdoctest: +SKIP("no rank")
+        >>> # Synchronous ops
+        >>> with _coalescing_manager():
+        >>>     for i in range(num_colls):
+        >>>         dist.all_reduce(tensors[i])
+        >>> # Asynchronous ops
+        >>> with _coalescing_manager(async_ops=True) as cm:
+        >>>     for i in range(num_colls):
+        >>>         dist.all_reduce(tensors[i])
+        >>> cm.wait()
+
+    .. warning::
+       :func:`_coalescing_manager` currently do not support coalescing
+       all-reduces with different reduce operators, e.g.  `ReduceOp.SUM` mixed
+       with `ReduceOp.PRODUCT`.
+    """
+    group = group or _get_default_group()
+    op_list = _world.pg_coalesce_state.setdefault(group, [])
+    if op_list:
+        raise RuntimeError("ProcessGroup has non-empty op list at the start of coalescing")
+    if device:
+        group._start_coalescing(device)
+    cm = _CoalescingManager()
     try:
-        yield
-    finally:
-        group._end_coalescing(device, reqs)
+        yield cm
+    except Exception:
+        # Re-throw exception caught by code inside the context manager
+        raise
+    else:
+        op_list = _world.pg_coalesce_state.pop(group)
+        if op_list:
+            # Collectives supporting "Fast Path" coalescing are captured.
+            # See implementation in corresponding collective APIs.
+            # Currently supported:
+            # - allreduce_coalesced
+            op0 = op_list[0].op
+            if op0 == all_reduce:
+                tensors = []
+                for op in op_list:
+                    tensors.append(op.tensor)
+                opts = AllreduceCoalescedOptions()
+                opts.reduceOp = op_list[0].redop
+                work = group.allreduce_coalesced(tensors, opts)
+            else:
+                raise AssertionError(
+                    f"Coalescing manager does not support fast-path coalescing of {op0}, "
```

Comment:
I don't get why can't we honor the coalescing request even if it would produce worse perf.

I rather see a warning along the lines of "unsupported mixing all_reduce with all_gather io the same coalescing group. Performance will be terrible."

Reply:
We can honor those. Just not in the `else` branch here :) 
For example, we can do the following today even with this PR:
```
with _coalescing_manager():
    dist.broadcast(tensor1, 0)
    dist.reduce(tensor2, 0)
```